Port terminal.js to react

larsu opened the pull request:
This makes terminal.js be a single component `UserTerminal`, which uses the default `Terminal` component with `cockpit.user`'s shell.

stefwalter commented:
Looks like the tests are finding a problem in the react component implementation.

dperpeet commented on pkg/systemd/terminal.jsx:
```diff
@@ -0,0 +1,101 @@
+    },
+
+    onTitleChanged: function (title) {
+        this.refs.title.textContent = title;
```
Wouldn't this be more react-like if the title were a part of the state? If the component gets re-rendered for some reason, the text will revert to the default `Terminal`.

dperpeet commented on pkg/systemd/terminal.jsx:
```diff
@@ -0,0 +1,101 @@
+    render: function () {
+        var terminal;
+        if (this.state.channel)
+            terminal = <components.Terminal ref="terminal"
```
enclose multiline jsx in `( )`

dperpeet commented on pkg/systemd/terminal.jsx:
```diff
@@ -0,0 +1,101 @@
+            terminal = <components.Terminal ref="terminal"
+                                            className="console"
+                                            channel={this.state.channel}
+                                            onTitleChanged={this.onTitleChanged} />;
```
Does jsx/react automatically bind `this.onTitleChanged` to `this`? The `this` in the called function might be wrong at runtime.

dperpeet commented on pkg/systemd/terminal.jsx:
```diff
@@ -0,0 +1,101 @@
+                                            channel={this.state.channel}
+                                            onTitleChanged={this.onTitleChanged} />;
+        else
+            terminal = <span>Loading...</span>;
```
Add a TODO here, this should use the EmptyState pattern.

dperpeet commented on pkg/systemd/terminal.jsx:
```diff
@@ -0,0 +1,101 @@
+        return (
+            <div className="panel panel-default console-container">
+                <div className="panel-heading">
+                    <tt id="terminal-title" ref="title">Terminal</tt>
```
I don't think these elements should have fixed ids. Thats what the refs are for.

dperpeet commented on pkg/systemd/terminal.jsx:
```diff
@@ -0,0 +1,101 @@
+            <div className="panel panel-default console-container">
+                <div className="panel-heading">
+                    <tt id="terminal-title" ref="title">Terminal</tt>
+                    <button id="terminal-reset"
```
I don't think these elements should have fixed ids. Thats what the refs are for.

dperpeet commented:
You still need to include the js bundle in the html file and the require the terminal component with requirejs: https://github.com/dperpeet/cockpit/commit/94eee78bf8e6cfac37f63fc08f78267ba916292e

dperpeet commented on pkg/systemd/terminal.html:
```diff
@@ -7,23 +7,10 @@
-    });
-</script>
+    <script type="text/javascript" src="../base1/cockpit-components.js"></script>
+    <script src="terminal.js"></script>
```
I think you're including this twice.
I recommend:
```
    <!-- debug:start -->
    <script src="terminal.js"></script>
    <!-- debug:end -->
```
The file might not even be present when installing via packages.

larsu commented:
Thanks for your comments!

> Wouldn't this be more react-like if the title were a part of the state? If the component gets re-rendered for some reason, the text will revert to the default `Terminal`.

Yes, totally! Changed.

> Does jsx/react automatically bind `this.onTitleChanged` to `this`? The `this` in the called function might be wrong at runtime.

Yes it does: https://facebook.github.io/react/docs/tutorial.html#callbacks-as-props

> I don't think these elements should have fixed ids. Thats what the refs are for.

Oh, yes of course. Fixed.

>  Add a TODO here, this should use the EmptyState pattern.

I'm not sure. The empty state pattern is not meant to be shown while something is loading, is it? I agree though that this should get something different than that ugly `span`.

stefwalter commented:
Tests are failing. In particular check-terminal and check-embed

larsu commented:
We've come to the conclusion that we have to include the react components in the `base1` bundle, because the system bundle includes `terminal.js`, which depends on the `Terminal` component. Require was thus trying to load `cockpit-component-terminal`, which isn't loaded until one of the pages pulling in `cockpit-components` was loaded. The other option would have been to remove the terminal from the system module. That would have given us another unbundled module, though.

@dperpeet, I know you were against bundling components into base1. What do you think about this?

stefwalter commented:
> I know you were against bundling components into base1. What do you think about this?

Not answering for @dperpeet ... but none the less. If we're doubling down on react, and components ... then it makes sense to include it in the default bundle. Less so for angular. But I've been meaning to consolidate the stuff we actually deliver in base1 (ie: the non-debug stuff) before we call in "stable".

stefwalter commented:
Would be nice to get this in for the 105 release.

dperpeet commented:
I'm a bit on the fence about including all the react components in the base bundle. I guess we can, but we'll have to be extra careful about what to include there.

dperpeet commented:
> We've come to the conclusion that we have to include the react components in the base1 bundle, because the system bundle includes terminal.js, which depends on the Terminal component. Require was thus trying to load cockpit-component-terminal, which isn't loaded until one of the pages pulling in cockpit-components was loaded. The other option would have been to remove the terminal from the system module. That would have given us another unbundled module, though.

I stated my opinion about bundling above. The issue you mentioned could have been solved by including the react components before the package (system) bundle, as I did with setroubleshoot.

dperpeet commented on pkg/playground/react-patterns.html:
```diff
@@ -6,7 +6,7 @@
     <link href="../base1/cockpit.css" type="text/css" rel="stylesheet">
     <script src="../base1/bundle.js"></script>
-    <script src="../base1/cockpit-components.js"></script>
+    <script src="../performance/bundle.js"></script>
```
why do we need the complete bundle here, and not just dialog-view?

stefwalter replied:
The end game here is to include the dialog in the base1 bundle right? If so, I think this is an ok interim measure.

larsu replied:
Also so that we don't load dialog-view.js when the bundle hasn't been loaded yet.

dperpeet commented on pkg/systemd/terminal.jsx:
```diff
@@ -0,0 +1,106 @@
+        },
+
+        onResetClick: function (event) {
+            // only reset if we had a channel before (cockpit.user is filled in)
```
This will trigger on a click with secondary mouse button as well, but it can be changed in a follow-up

larsu replied:
Right. Thanks for letting me know. I've fixed it in the latest version.

dperpeet commented:
Looks good aside from the two comments / questions.
